snafu/benchmarks/uperf/uperf.py

```diff
@@ -0,0 +1,301 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""Wrapper for running the uperf benchmark. See http://uperf.org/ for more information."""
+from typing import Any, Dict, Iterable, List, Optional, Tuple, Union
+import re
+import datetime
+import dataclasses
+import shlex
+from snafu.benchmarks import Benchmark, BenchmarkResult
+from snafu.config import Config, ConfigArgument, FuncAction, check_file, none_or_type
+from snafu.process import sample_process, ProcessSample
+
+
+class ParseRangeAction(FuncAction):
+    """Parses node_range and density_range attributes."""
+
+    def func(self, arg: str) -> List[int]:
+        return [int(x) for x in arg.split("-") if x != ""]
+
+
+@dataclasses.dataclass
+class RawUperfStat:
+    timestamp: float
+    bytes: int
+    ops: int
+
+
+@dataclasses.dataclass
+class UperfStdout:
+    results: Tuple[RawUperfStat, ...]
+    duration: int
+    test_type: Optional[str] = None
+    protocol: Optional[str] = None
+    message_size: Optional[int] = None
+    read_message_size: Optional[int] = None
+    num_threads: Optional[int] = None
+
+
+@dataclasses.dataclass
+class UperfConfig:
+    test_type: Optional[str] = None
+    protocol: Optional[str] = None
+    message_size: Optional[int] = None
+    read_message_size: Optional[int] = None
+    num_threads: Optional[int] = None
+    duration: Optional[int] = None
+    kind: Optional[str] = None
+    hostnetwork: Optional[str] = None
+    remote_ip: Optional[str] = None
+    client_ips: Optional[str] = None
+    service_ip: Optional[str] = None
+    client_node: Optional[str] = None
+    server_node: Optional[str] = None
+    num_pairs: Optional[str] = None
+    multus_client: Optional[str] = None
+    networkpolicy: Optional[str] = None
+    density: Optional[str] = None
+    nodes_in_iter: Optional[str] = None
+    step_size: Optional[str] = None
+    colocate: Optional[str] = None
+    density_range: Optional[List[int]] = None
+    node_range: Optional[List[int]] = None
+    pod_id: Optional[str] = None
+
+    @classmethod
+    def new(cls, stdout: UperfStdout, config: Config):
+        kwargs: Dict[str, Any] = dict()
+        for fields in dataclasses.fields(cls):
+            val = getattr(stdout, fields.name, None)
+            if val is None:
+                val = getattr(config, fields.name, None)
+            kwargs[fields.name] = val
+        return cls(**kwargs)
+
+
+@dataclasses.dataclass
+class UperfStat:
+    uperf_ts: str
+    bytes: int
+    norm_byte: int
+    ops: int
+    norm_ops: int
+    norm_ltcy: float
+    iteration: Optional[int] = None
+
+
+class Uperf(Benchmark):
+    """
+    Wrapper for the uperf benchmark.
+    """
+
+    tool_name = "uperf"
+    args = (
```

Comment:
Hey @learnitall, small nit. Why do we need to define these flags as ConfigArgument?